### PR TITLE
Loading the configuration outside of the functions #67

### PR DESCRIPTION
**Pull request by bazkiaei**
The goal of this PR is to make the code not to load the configuration more than once.

### CODE REVIEW

File: mocks/mocks.py
Comment (bazkiaei) [inline mocks/mocks.py]: Based on Anthony's advice, I tried to create a dictionary and pass it through other functions instead of kwargs.
So the user should load the configuration first and pass it through other functions. 
I hope this is what I was meant to do.

File: mocks/tests/conftest.py
Comment (bazkiaei) [inline mocks/tests/conftest.py]: I did not want to load the configuration file for tests other than the one which is testing the `mocks.load_configuration` so I created this `fixture`. It is too long and that makes me think maybe I should try something else, so I will appreciate any advice about it.

**Comment by bazkiaei** [comment]
@lspitler @AnthonyHorton @wtgee 
Ready for the review.

File: mocks/mocks.py
Comment (AnthonyHorton) [reply mocks/mocks.py]: This function is pointless, the effect of having this is just to create an alternative name for `mocks.utils.yaml_load_config`. You could achieve exactly the same effect with `from mock.utils import load_yaml_config as load_configuration`, but there's no advantage in doing that, either.

What I suggested was creating a main function that would load the configuration _and then call all of the other functions_. That would be useful, because then the user would be able to create simulated images from simulation data without having to work out what order to call all of these functions in, or write their own code to so.

The reason why we originally suggesting splitting up the one big function you started with into a series of smaller, self contained functions wasn't that we thought that there shouldn't be a function that does the whole thing. The reason for splitting it up into multiple functions was to make the code easier to understand, test, maintain and reuse. There should still be a function that does the whole thing, it's just that instead of a great big block of code it should consist mostly of calls to the other functions.

File: mocks/mocks.py
Comment (AnthonyHorton) [inline mocks/mocks.py]: Usually? Seems like you have to call `prepare_mocks` on the `config` before calling this otherwise it won't work.

File: mocks/tests/conftest.py
Comment (AnthonyHorton) [reply mocks/tests/conftest.py]: Why not use configuration files for all the tests? It would be simpler than this.

What you do is have a test fixture that loads the config and returns it (either here or in `tests/test_mocks.py`), then have a test use the fixture and check that it loaded OK, e.g.

```
from mock.utils import load_yaml_config

@pytest.fixture(scope='package')
def configuration():
    return load_yaml_config()
```

and

```
def test_load_configuration(configuration):
    assert configuration['data_path'] == 'sim_data/cl19.fits'
    assert configuration['galaxy_distance'] == 10.
￼    assert isinstance(configuration, dict)
```
The first test to use the `configuration` fixture will trigger the call to `load_yaml_config`, and if that fails with an exception then the exception will be passed up to the test, and it will fail (as it should). If it passes then the rest of the tests can run using the same `configuration` fixture to access the test config.

File: mocks/tests/test_mocks.py
Comment (AnthonyHorton) [inline mocks/tests/test_mocks.py]: Generally speaking `isinstance(object, class)` is preferred for testing the type of an object, not `type(object) == class`. Several reasons for that, the main one being that `isinstance` allows for subclasses.

**Review by AnthonyHorton — CHANGES_REQUESTED** [review]
OK, seems like I didn't successful explain my suggestion about where to load the configuration. See comments.

File: mocks/mocks.py
Comment (bazkiaei) [reply mocks/mocks.py]: There are some inconsistencies among the functions of `mocks.py`. For instance functions `scale_light_by_distance` and `prepare_mocks` both use the distance to the galaxy to scale its light. 
The first step to create the main function is making all functions consistent.

Comment (bazkiaei) [reply mocks/mocks.py]: After making the functions consistent, the jupyter-notebook examples should be updated as the base for the main function that calls other functions and does all the steps to create the mock image.
These notebooks could be used to create the main function.

File: mocks/tests/conftest.py
Comment (bazkiaei) [reply mocks/tests/conftest.py]: I applied your suggestion, but is that alright that the function `utils.load_yaml_config` is tested twice? Once here and another time in `test_utils.py`

File: mocks/tests/conftest.py
Comment (bazkiaei) [inline mocks/tests/conftest.py]: When the fixture scope was `package` and I run the test locally there were no errors and it was totally alright but Travis raised an error for that!

Comment (AnthonyHorton) [reply mocks/tests/conftest.py]: OK, package scope was new in pytest 3.7, Travis is probably using an older version by default. For now just use module scope.

**Comment by bazkiaei** [comment]
@AnthonyHorton 
Ready for review.

File: mocks/mocks.py
Comment (AnthonyHorton) [inline mocks/mocks.py]: Physical distance is (or at least can be) in the config, instead you're using a separate argument for it and ignoring the config.  You should be getting this value from `config` instead, and using the `config` parsing code to make sure the correct value of `physical_distance` is already in there (i.e. unit conversion, applying default value if not in the file, or overriding config file value with another).

File: mocks/mocks.py
Comment (AnthonyHorton) [inline mocks/mocks.py]: I think `target_galaxy_comoving_depth` is also a value that can be in the config, so as with `physical_distance` it should be. Part of the benefit of passing around a `config` dictionary is that you don't need lots of individual arguments for all your functions.

File: mocks/mocks.py
Comment (AnthonyHorton) [inline mocks/mocks.py]: This function is doing 4 unrelated things. I would suggest splitting this into at least three functions.

The first function fully parses the config, i.e. loads the config file with `mocks.utils.load_yaml_config`, then goes through _all_ the parameters (not just `galaxy_coordinates` & `observation_time` as now) and applies any keyword arguments that override the config file values, gets the remaining values from the config file, adds default values for any parameters that are missing, and does any required unit conversions, etc.  This function returns the fully populated, validated config dict for all the other functions to use.  The other functions then don't have to do any fiddling around with overriding values, default values, unit conversions, etc.

The second function takes the config as its only argument, loads the simulation data, and returns it. This would be the place to do 32 bit to 64 bit float type conversion, when required.

The calls to `compute_total_mass` and `compute_apparent_ABmag` also don't belong here.  You could create a 3rd function that calls them, using values from `config`,  and returns the results, or you could do those calls in the main function.

The general principle here is that a function should do one clearly defined thing, the thing that the function name implies that it does. Adding other, unrelated things to a function makes it much harder to understand the software.

**Review by AnthonyHorton — CHANGES_REQUESTED** [review]
A good step in right direction, but use of config versus function arguments is still patchy, and you have parsing of parameters spread around several functions instead of in one `parse_config` function.

File: mocks/mocks.py
Comment (AnthonyHorton) [inline mocks/mocks.py]: Don't do all this stuff in this function, it doesn't belong here. Do it in a `parse_config` function.

Comment (bazkiaei) [reply mocks/mocks.py]: @AnthonyHorton 
Yes, I will. I just referenced your previous comment in another issue to start creating some functions, including `parse_config`, in another PR.

Comment (bazkiaei) [reply mocks/mocks.py]: This is just a clue for me for creating the `parse_config` function.

Comment (AnthonyHorton) [reply mocks/mocks.py]: Why not just put it in the right place to begin with? It's not off topic for this PR, which is about improving config loading. Seems to just be creating more work to write the code in one place in one PR, then immediately move it somewhere else in another one. You'll have to modify a bunch of tests twice if you do that.

File: config_directory/config_example_old.yaml
Comment (bazkiaei) [inline config_directory/config_example_old.yaml]: After creating the `parse_config` function, more cosmological parameters are added to `config_examle.yaml` and since the new function should be tested in different ways, the added cosmological parameters are different to the default cosmology, `WMAP9`. These different parameters cause changes in the result of many functions in `test_mocks.py`. To avoid huge changes in the tests, I added this file with the same values that `config_example.yaml` used to have.
However, I am not very happy with the name that I have chosen for it, maybe `config_old_example.yaml` makes more sense.

File: config_directory/config_no_cosmology_example.yaml
Comment (bazkiaei) [inline config_directory/config_no_cosmology_example.yaml]: Oh, I forgot it again!

File: config_directory/config_no_cosmology_example.yaml
Comment (bazkiaei) [inline config_directory/config_no_cosmology_example.yaml]: This file is added for testing the `parse_config` file when there is no cosmological parameter in the configuration file provided. The function is supposed to use the default cosmology in that case.

File: config_directory/config_none_value_example.yaml
Comment (bazkiaei) [inline config_directory/config_none_value_example.yaml]: 🔢

File: config_directory/config_none_value_example.yaml
Comment (bazkiaei) [inline config_directory/config_none_value_example.yaml]: This file has been added to test the capability of the `parse_config` function when there are defined cosmological parameters in the configuration file but they have no value.

File: mocks/mocks.py
Comment (bazkiaei) [inline mocks/mocks.py:102]: For cosmological parameters, the code first tries to get the value from the `kwargs` and if it does not exist, the code will try the `config`. If the key does not exist in the `config`, it will except the `KeyError` and set the value by default cosmology.
The second `try .... expext .... ` works when the parameter is defined in the configuration file but has no value. This block will set the value using the default cosmology and prevent the existence of cosmological parameters with value `None`.
I have not done the same for keys other than the cosmological parameters because I thought an error should raise when they have the value `None`, but I am not sure if the error should be raised here in the `parse_config`.

**Review by bazkiaei — COMMENTED** [review]
@AnthonyHorton 
The function `parse_config` and its tests are created. But it is not used by any other function yet.

File: mocks/mocks.py
Comment (bazkiaei) [inline mocks/mocks.py]: To make the code works for the Dark Matter only simulations, I think maybe I should change the lines that are parsing baryonic matter density as follow:
```
    try:
        config['Omega_b'] = kwargs.get('Omega_b',
                                       config['Omega_b'])
    except KeyError:
        config['Omega_b'] = default_cosmology.Ob0
    try:
        assert config['Omega_b']
    except AssertionError:
        if config['Omega_b'] is not 0:
            config['Omega_b'] = default_cosmology.Ob0
```

File: config_directory/config_example.yaml
Comment (AnthonyHorton) [inline config_directory/config_example.yaml]: Might include a comment about which cosmology these example parameters correspond too, e.g. WMAP9.

File: config_directory/config_none_value_example.yaml
Comment (AnthonyHorton) [reply config_directory/config_none_value_example.yaml]: I approve of the thoroughness, but I'm not sure this is entirely necessary. Seems like a strange thing for the user to do.

File: mocks/mocks.py
Comment (AnthonyHorton) [inline mocks/mocks.py]: Technically this should be `astropy.cosmology.Cosmology`. That's the base class for astropy cosmologies. `astropy.cosmology` is the module that contains the cosmology stuff.

File: mocks/mocks.py
Comment (AnthonyHorton) [inline mocks/mocks.py]: Do you want default values for any of these parameters, in case they aren't in either the config file or `kwargs`? Can do this using the following pattern:

```
config[<parameter_name>] = kwargs.get(<parameter_name>,
                                       config.get(<parameter_name>,
                                                  <default value>))
```

Might make sense to do this for parameters you don't expect to be changed by the user very often.

File: mocks/mocks.py
Comment (AnthonyHorton) [reply mocks/mocks.py:102]: See comment above about a neater way to do this. In this case it would become:
```
config['hubble_constant'] = kwargs.get('hubble_constant',
                                        config.get('hubble_constant',
                                                    default_cosmology.H0))
```

I don't think you need to worry about the case where the key is in the config file but has no value. I think it's fair to just consider that a bad config file.  If the user wants to revert to using default values without removing the key from the config file they should just comment the line, not remove the value.

File: mocks/mocks.py
Comment (AnthonyHorton) [reply mocks/mocks.py]: As a general point, if you want to distinguish between something being `None`, or `0`, or some other value which evaluates to `False` (e.g. an empty `list`) then you can test for those directly, e.g. `if config['Omega_b'] is None:` or `if config['Omega_b'] == 0:`, etc.

As I noted in comments above I really don't think you need to allow for keys without values in the config file so you don't need this added complexity at all.

File: mocks/mocks.py
Comment (AnthonyHorton) [inline mocks/mocks.py]: Don't need `**kwargs` now. Overriding config values should now be done in the call to `parse_config()`, and _only_ in `parse_config()`. Otherwise there's a risk of accidentally using inconsistent config values at different stages in the calculation.

File: mocks/mocks.py
Comment (AnthonyHorton) [inline mocks/mocks.py]: This isn't an argument of this function anymore. Should remove from docstring.

File: mocks/mocks.py
Comment (AnthonyHorton) [reply mocks/mocks.py]: This is all redundant now that it's being done in `parse_config`. You should remove it from here.

File: mocks/mocks.py
Comment (AnthonyHorton) [inline mocks/mocks.py]: Should not be able to. As in comment above, the only place to override configuration parameters should be `parse_config()` in order to ensure a consistent set of config parameters are used by all the functions.

File: mocks/mocks.py
Comment (AnthonyHorton) [inline mocks/mocks.py]: Remove this.

File: mocks/tests/test_mocks.py
Comment (AnthonyHorton) [inline mocks/tests/test_mocks.py]: As notes in other comments, I don't think you need to allow for the possibility of keys without values in the config file. That's just a bad config file.

**Review by AnthonyHorton — CHANGES_REQUESTED** [review]
Part of the reason for parsing the configuration in a dedicated function is to ensure all the other functions use a consistent set of configuration parameters. Consequently you should remove the option to override configuration parameters with keyword argument from the other functions. `parse_config` should be only place to do that.

Also, nested `.get()` is a better way to implement the 'keyword arg, or config file, or default value' pattern than the forced exceptions you are using.

You can also simplify things by dropping the requirement to be able to have keys in the config file with no value. I don't know why you would need to be able to do that.

File: mocks/mocks.py
Comment (bazkiaei) [reply mocks/mocks.py]: Should I choose any particular coordinate as the default or just something like what exists in the config file would work?

File: config_directory/config_example.yaml
Comment (bazkiaei) [reply config_directory/config_example.yaml]: Some of them are the parameters that have been used in the simulation and I just have chosen the others randomly. I intentionally wanted them to be different from WMAP9 for tests purposes.

**Comment by bazkiaei** [comment]
@AnthonyHorton 
I tried to apply your advice. I had a meeting with @lspitler and he advised me to close this PR when `parse_config` is ready and start other new functions in other PRs. So if you agree with that and you are satisfied with this function I can start creating the other functions in other PRs, one by one.

File: mocks/mocks.py
Comment (lspitler) [reply mocks/mocks.py:102]: @bazkiaei I don't think we need to worry about the case where they don't provide part of the cosmological parameters, since the mock code is no longer doing what the user wants - converting a simulation into something realistic. SImilarly, there is no case where we will send in cosmological data that should default to a cosmology that you've selected here - *all* cosmological simulations will have been run with a cosmology. If that cosmology is unknown, then the simulation is worthless to us. If that cosmology is known, then we can just require the user to provide the cosmology exactly.

To summarise, there is no use case to build in a flexible or default cosmology as the cosmological data will always have a cosmology that is known perfectly.

File: config_directory/config_no_cosmology_example.yaml
Comment (lspitler) [reply config_directory/config_no_cosmology_example.yaml]: I don't think this is needed given my comments below.

File: config_directory/config_example_old.yaml
Comment (lspitler) [reply config_directory/config_example_old.yaml]: I don't think this is needed given my comments below.

File: config_directory/config_example.yaml
Comment (lspitler) [reply config_directory/config_example.yaml]: > Different from WMAP9 for tests purposes

that sounds like a test of `astropy.cosmology` not `huntsman-mocks`, which shouldn't care about what cosmology is needed.

Are you sure you need this test?

File: mocks/mocks.py
Comment (lspitler) [inline mocks/mocks.py]: The prior three lines do nothing. delete them.

File: mocks/mocks.py
Comment (lspitler) [inline mocks/mocks.py:250]: these should be removed I think

File: mocks/mocks.py
Comment (lspitler) [inline mocks/mocks.py]: I think the old way is preferred: `from mocks.utils import load_yaml_config` since it only imports the function that is needed.

**Review by lspitler — CHANGES_REQUESTED** [review]
Some comments given. In retrospect, the discussion about cosmology and related tests might have been better suited for a separate PR as its separate to the way the config is loaded. But it seems it has merged here, so let's treat this PR as both finalising the expected cosmology and dealing with how to load the config.

**Comment by bazkiaei** [comment]
@lspitler 
I tried to apply your comments. I used WMAP9 for the configuration file and canceled the default cosmology from `parse_config`. Ready to review.
@AnthonyHorton

File: config_directory/config_example.yaml
Comment (lspitler) [inline config_directory/config_example.yaml]:
```suggestion
# Cosmological parameters, WMAP9 as default. These should map onto astropy.cosmology
```

File: mocks/tests/conftest.py
Comment (lspitler) [inline mocks/tests/conftest.py]: why not just use `config_example.yaml`? why do you need old?

File: mocks/mocks.py
Comment (lspitler) [inline mocks/mocks.py:26]: maybe this should move into `utils`?

File: mocks/mocks.py
Comment (lspitler) [inline mocks/mocks.py]: remove this

File: mocks/mocks.py
Comment (lspitler) [reply mocks/mocks.py:250]: I see you resolved this, which I guess means its OK to leave these. Let me know in slack why you include Deleted Parameters -- I've never heard of this and want to learn.

**Review by lspitler — APPROVED** [review]
Looks good @bazkiaei I think this is ready after you address my remaining comments.

File: mocks/tests/conftest.py
Comment (bazkiaei) [reply mocks/tests/conftest.py]: There will be just `config_example.yamle` as the configuration file and I will remove `config_example_old.yaml`. I will do that while I am creating the jupyter notebook examples. 
At that stage, there is more consistency among functions and I will try to go through all the code.

File: mocks/mocks.py
Comment (bazkiaei) [reply mocks/mocks.py:26]: I think we can use some advice from @AnthonyHorton  and @wtgee about that.

Comment (lspitler) [reply mocks/mocks.py:26]: @bazkiaei I like this definition:

> Looks to me like /utils is a place where you can place small snippets you can use throughout the application. Small functions to build bigger things with.

from https://github.com/erikras/react-redux-universal-hot-example/issues/808 

You designed `parse_config` to be used specifically for `mocks.py`, so I guess it makes sense. 

This is scope creep on the PR, so let's close this PR without changing this, but create a new Issue to decide where *all* `mocks.py` functions should go -- other ones I think meet the definition I copied in above. Or better yet add it to the discussion about creating a Class for Mocks: #63

File: mocks/tests/conftest.py
Comment (lspitler) [reply mocks/tests/conftest.py]: @bazkiaei its more appropriate to deal with this here I think than in a PR about an example notebook. Isn't it just changing the filename? What's different about `_old.yaml` that is needed right now that impacts everything?

**Comment by lspitler** [comment]
@AnthonyHorton your request for review is blocking this PR - can you Approve when you get a chance? I'm happy with @bazkiaei  changes based upon your comments.

**Review by AnthonyHorton — APPROVED** [review]

**Comment by AnthonyHorton** [comment]
OK, looks good now. LGTM.

**Pull request merged**
